pointercrate-user/src/auth/get.rs

```diff
@@ -46,9 +60,117 @@ impl AuthenticatedUser {
+        // validate a fresh token, as it is guaranteed to be valid.
+        //
+        // https://developers.google.com/identity/openid-connect/openid-connect#obtainuserinfo
+        let key = DecodingKey::from_secret(&[]);
+        let mut validation = Validation::new(Algorithm::HS256);
+        validation.insecure_disable_signature_validation();
+        validation.validate_exp = false;
+        validation.validate_nbf = false;
+        validation.validate_aud = false;
+
+        let user_info = jsonwebtoken::decode::<GoogleUserInfo>(&response.id_token, &key, &validation)
+            .map_err(|_| CoreError::Unauthorized)?
+            .claims;
+
+        match User::by_google_account(&user_info.id, connection).await {
+            Ok(user) => Ok(AuthenticatedUser {
+                user,
+                password_hash: None,
+                google_account_id: Some(user_info.id),
+                email_address: Some(user_info.email),
+            }),
+            Err(UserError::UserNotFoundGoogleAccount { .. }) => {
+                if let Some(id) = existing_id {
+                    let user = AuthenticatedUser::by_id(id, connection).await?;
+
+                    if user.google_account_id.is_some() {
+                        return Err(CoreError::Unauthorized.into());
+                    }
+
+                    let updated_user = sqlx::query!(
+                        "UPDATE members SET google_account_id = $1, email_address = ($2::text)::email WHERE member_id = $3 RETURNING member_id",
+                        user_info.id,
+                        user_info.email,
+                        id
+                    )
+                    .fetch_one(connection)
+                    .await?;
+
+                    if updated_user.member_id != id {
+                        return Err(Error::RowNotFound.into());
+                    }
+
+                    Ok(AuthenticatedUser {
+                        user: User { id, ..user.user },
+                        password_hash: None,
+                        google_account_id: Some(user_info.id),
+                        email_address: Some(user_info.email),
+                    })
+                } else {
+                    // This will never conflict with an existing user
+                    // According to Google, the account ID is always unique
+                    // https://developers.google.com/identity/openid-connect/openid-connect#an-id-tokens-payload
+                    let name = format!("{}#{}", user_info.name, user_info.id);
+
+                    let id = sqlx::query!(
+                        "INSERT INTO
+                            members (email_address, name, display_name, google_account_id)
+                        VALUES
+                            (($1::text)::email, $2, $3, $4) RETURNING member_id
+                        ",
+                        user_info.email,
+                        name,
```

Comment:
nit: maybe do the concatenation in the query?
```suggestion
                    // This will never conflict with an existing user
                    // According to Google, the account ID is always unique
                    // https://developers.google.com/identity/openid-connect/openid-connect#an-id-tokens-payload
                    let id = sqlx::query!(
                        "INSERT INTO
                            members (email_address, name, display_name, google_account_id)
                        VALUES
                            (($1::text)::email, $2 || $3, $4, $5) RETURNING member_id
                        ",
                        user_info.email,
                        user_info.name,
                        user_info.id,
```